src/net/ossl.cc

```diff
@@ -717,6 +722,23 @@ class session : public enable_shared_from_this<session>, public session_impl {
         return make_ready_future<>();
     }
 
+    // Call this method when its desired to have a read performed, but if a read is already being
+    // performed, just return when that read completes
+    future<> wait_for_data() {
+        auto avail = BIO_pending(_in_bio);
+        auto units = try_get_units(_in_sem, 1);
+        if (units) {
+            units->return_all();
+            return wait_for_input();
+        }
```

Comment:
I'm a little confused by this.  If you get units, you immediately return then and then wait on the input (`_in`) stream.  Why don't you reutrn the units after the return from wait_for_input?

Reply:
Nice catch, this wasn't correctly updated after the latest force push

Reply:
On another note, not really sure how to test this edge case. It occurs when `SSL_write` gets `SSL_ERROR_WANT_READ`